netkan/netkan/download_counter.py

```diff
@@ -0,0 +1,114 @@
+from json.decoder import JSONDecodeError
+
+
+class DownloadCounter:
+
+    def __init__(self, netkan_path, ckan_meta, github_token):
+        self.netkan_path = Path(netkan_path)
+        self.ckan_meta = ckan_meta
+        self.github_headers = {'Authorization': f'token {github_token}'}
+        self.kref_pattern = re.compile('^#/ckan/([^/]+)/(.+)$')
+        self.github_pattern = re.compile('^([^/]+)/([^/]+)')
+        self.counts = {}
+        self.output_file = Path(
+            self.ckan_meta.working_dir, 'download_counts.json'
+        )
+
+    def count_from_spacedock(self, mod_id):
+        return requests.get(
+            f'https://spacedock.info/api/mod/{mod_id}'
+        ).json()['downloads']
+
+    def count_from_github(self, mod_id):
+        (user, repo) = self.github_pattern.match(mod_id).groups()
+        releases = requests.get(
+            f'https://api.github.com/repos/{user}/{repo}/releases',
+            headers=self.github_headers
+        ).json()
+
+        total = 0
+        for rel in releases:
+            for asset in rel['assets']:
+                total += asset['download_count']
+
+        repo = requests.get(
+            f'https://api.github.com/repos/{user}/{repo}',
+            headers=self.github_headers
+        ).json()
+        if 'parent' in repo:
+            total += self.count_from_github(repo['parent']['full_name'])
+        return total
+
+    def count_from_curse(self, mod_id):
+        return requests.get(
+            f'https://api.cfwidget.com/project/{mod_id}'
+            if mod_id.isnumeric()
+            else f'https://api.cfwidget.com/kerbal/ksp-mods/{mod_id}'
+        ).json()['downloads']['total']
+
+    def count_from_url(self, url):
+        count = 0
+        try:
+            count = self.count_from_netkan(requests.get(url).json())
+        except JSONDecodeError:
+            logging.error("Failed to get count from {}".format(url))
+        return count
+
+    def count_from_netkan(self, netkan):
+        kref = netkan.get('$kref', None)
+        if not kref:
+            return 0
+        (kind, mod_id) = self.kref_pattern.match(kref).groups()
+        if kind == 'netkan':
+            return self.count_from_url(mod_id)
+        elif kind == 'spacedock':
+            return self.count_from_spacedock(mod_id)
+        elif kind == 'github':
+            return self.count_from_github(mod_id)
+        elif kind == 'curse':
+            return self.count_from_curse(mod_id)
+        return 0
+
+    def get_counts(self):
+        for netkan in self.netkan_path.glob('NetKAN/*.netkan'):
+            netkan = Path(netkan)
+            count = self.count_from_netkan(
+                json.loads(netkan.read_text())
+            )
+            if count > 0:
+                self.counts[netkan.stem] = count
+
+    def write_json(self):
+        self.output_file.write_text(
+            json.dumps(self.counts, sort_keys=True, indent=4)
+        )
+
+    def counts_changed(self):
```

Comment:
I think this isn't called from anywhere. Looks like it should be used by `update_counts` to decide whether to call `commit_counts`?

Reply:
Oh gosh you're totally right, good catch! I was definitely getting tired when I finished this up.

GitPython will commit even when there are no changes, so you have to do those checks yourself.

I think the author intended it to be a low level API consumed by a project to put a friendly wrapper around it. Which comes with side effects of having to deal with the bits the git client normally takes care of for you.

I'll sort this out to, thank you!